File: backend/src/api/router/llm/technicians.py
"""
Technicians API Router
CRUD operations for technicians database
"""

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
import sys
import os

# Add the project root so we can import src
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', '..', '..')))
from src.services.mcp.mcp_technician_service import TechnicianService
from src.core.logger import log

logger = log.bind(name="api.router.llm.technicians")

router = APIRouter()

# Initialize service
technician_service = TechnicianService()


# ==================== SCHEMAS ====================

class Certification(BaseModel):
    certification_type: str
    certification_name: str
    issue_date: Optional[str] = None
    expiry_date: Optional[str] = None


class TechnicianBase(BaseModel):
    first_name: str
    last_name: str
    phone: Optional[str] = None
    email: Optional[str] = None
    company: Optional[str] = None
    certifications: List[Certification] = []
    metadata: Dict[str, Any] = {}


class TechnicianResponse(TechnicianBase):
    id: str
    created_at: str
    updated_at: str


class TechnicianUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    company: Optional[str] = None
    certifications: Optional[List[Certification]] = None
    metadata: Optional[Dict[str, Any]] = None


# ==================== ROUTES ====================

@router.get("/", response_model=Dict[str, Any])
async def list_technicians():
    """Get all technicians"""
    try:
        technicians = technician_service.get_all_technicians()
        return {
            "success": True,
            "data": technicians,
            "total": len(technicians)
        }
    except Exception as e:
        logger.error(f"Error loading technicians: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to load technicians"
        )


@router.get("/{tech_id}", response_model=Dict[str, Any])
async def get_technician(tech_id: str):
    """Get a single technician by ID"""
    try:
        technician = technician_service.get_technician_by_id(tech_id)
        
        if not technician:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Technician not found"
            )
        
        return {
            "success": True,
            "data": technician
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error loading technician: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to load technician"
        )


@router.post("/", response_model=Dict[str, Any], status_code=status.HTTP_201_CREATED)
async def create_technician(technician: TechnicianBase):
    """Create a new technician"""
    try:
        # Convert Pydantic model to dict
        tech_dict = technician.dict()
        
        # Convert certifications to dicts
        tech_dict['certifications'] = [cert.dict() for cert in technician.certifications]
        
        # Upsert (will add new if doesn't exist)
        result = technician_service.upsert_technician(tech_dict)
        
        if result['action'] != 'added':
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Technician with this name already exists"
            )
        
        return {
            "success": True,
            "data": result['technician']
        }
    except HTTPException:
        raise
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        logger.error(f"Error creating technician: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create technician"
        )


@router.put("/{tech_id}", response_model=Dict[str, Any])
async def update_technician(tech_id: str, updates: TechnicianUpdate):
    """Update an existing technician"""
    try:
        # Convert to dict and remove None values
        update_dict = {k: v for k, v in updates.dict().items() if v is not None}
        
        # Convert certifications to dicts if present
        if 'certifications' in update_dict:
            update_dict['certifications'] = [cert.dict() for cert in updates.certifications]
        
        updated_tech = technician_service.update_technician(tech_id, update_dict)
        
        if not updated_tech:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Technician not found"
            )
        
        return {
            "success": True,
            "data": updated_tech
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating technician: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update technician"
        )


@router.delete("/{tech_id}", response_model=Dict[str, Any])
async def delete_technician(tech_id: str):
    """Delete a technician"""
    try:
        deleted = technician_service.delete_technician(tech_id)
        
        if not deleted:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Technician not found"
            )
        
        return {
            "success": True,
            "message": "Technician deleted successfully"
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting technician: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete technician"
        )

File: frontend/routes/technicians.ts
import { Request, Response, Router } from 'express';
import fs from 'fs/promises';
import path from 'path';

const router = Router();

// Path to technicians database
const TECHNICIANS_DB_PATH = path.resolve(__dirname, '../../mcp/technicians_database.json');

interface Technician {
  id: string;
  first_name: string;
  last_name: string;
  phone: string | null;
  email: string | null;
  company: string | null;
  certifications: Array<{
    certification_type: string;
    certification_name: string;
    issue_date: string | null;
    expiry_date: string | null;
  }>;
  created_at: string;
  updated_at: string;
  metadata?: any;
}

interface TechniciansDatabase {
  technicians: Technician[];
  metadata?: {
    created: string;
    last_updated: string;
    version: string;
  };
}

// Helper to load database
async function loadDatabase(): Promise<TechniciansDatabase> {
  try {
    const data = await fs.readFile(TECHNICIANS_DB_PATH, 'utf-8');
    return JSON.parse(data);
  } catch (error) {
    // Return empty structure if file doesn't exist
    return {
      technicians: [],
      metadata: {
        created: new Date().toISOString(),
        last_updated: new Date().toISOString(),
        version: '1.0.0'
      }
    };
  }
}

// Helper to save database
async function saveDatabase(db: TechniciansDatabase): Promise<void> {
  if (db.metadata) {
    db.metadata.last_updated = new Date().toISOString();
  }
  await fs.mkdir(path.dirname(TECHNICIANS_DB_PATH), { recursive: true });
  await fs.writeFile(TECHNICIANS_DB_PATH, JSON.stringify(db, null, 2), 'utf-8');
}

// GET /api/technicians - Get all technicians
router.get('/', async (req: Request, res: Response) => {
  try {
    const db = await loadDatabase();
    res.json({
      success: true,
      data: db.technicians,
      total: db.technicians.length
    });
  } catch (error) {
    console.error('Error loading technicians:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to load technicians'
    });
  }
});

// GET /api/technicians/:id - Get single technician
router.get('/:id', async (req: Request, res: Response) => {
  try {
    const db = await loadDatabase();
    const technician = db.technicians.find(t => t.id === req.params.id);
    
    if (!technician) {
      res.status(404).json({
        success: false,
        error: 'Technician not found'
      });
      return;
    }
    
    res.json({
      success: true,
      data: technician
    });
  } catch (error) {
    console.error('Error loading technician:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to load technician'
    });
  }
});

// PUT /api/technicians/:id - Update technician
router.put('/:id', async (req: Request, res: Response) => {
  try {
    const db = await loadDatabase();
    const index = db.technicians.findIndex(t => t.id === req.params.id);
    
    if (index === -1) {
      res.status(404).json({
        success: false,
        error: 'Technician not found'
      });
      return;
    }
    
    const existingTech = db.technicians[index];
    const updatedTech: Technician = {
      ...existingTech,
      ...req.body,
      id: existingTech.id, // Preserve ID
      created_at: existingTech.created_at, // Preserve creation date
      updated_at: new Date().toISOString()
    };
    
    db.technicians[index] = updatedTech;
    await saveDatabase(db);
    
    res.json({
      success: true,
      data: updatedTech
    });
  } catch (error) {
    console.error('Error updating technician:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update technician'
    });
  }
});

// DELETE /api/technicians/:id - Delete technician
router.delete('/:id', async (req: Request, res: Response) => {
  try {
    const db = await loadDatabase();
    const index = db.technicians.findIndex(t => t.id === req.params.id);
    
    if (index === -1) {
      res.status(404).json({
        success: false,
        error: 'Technician not found'
      });
      return;
    }
    
    db.technicians.splice(index, 1);
    await saveDatabase(db);
    
    res.json({
      success: true,
      message: 'Technician deleted successfully'
    });
  } catch (error) {
    console.error('Error deleting technician:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete technician'
    });
  }
});

// POST /api/technicians - Create new technician
router.post('/', async (req: Request, res: Response) => {
  try {
    const db = await loadDatabase();
    
    // Generate ID from name
    const id = `${req.body.first_name}_${req.body.last_name}`.toLowerCase().replace(/\s+/g, '_');
    
    // Check if already exists
    if (db.technicians.some(t => t.id === id)) {
      res.status(409).json({
        success: false,
        error: 'Technician with this name already exists'
      });
      return;
    }
    
    const newTech: Technician = {
      id,
      first_name: req.body.first_name,
      last_name: req.body.last_name,
      phone: req.body.phone || null,
      email: req.body.email || null,
      company: req.body.company || null,
      certifications: req.body.certifications || [],
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      metadata: req.body.metadata || {}
    };
    
    db.technicians.push(newTech);
    await saveDatabase(db);
    
    res.status(201).json({
      success: true,
      data: newTech
    });
  } catch (error) {
    console.error('Error creating technician:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to create technician'
    });
  }
});

export default router;
